### PR TITLE
[moslo] Recursive library dependency search with objdump. Fixes JB#28364

### PR DESCRIPTION
**Pull request by kjokinie**
The objdump based search was not working correctly as ld-linux.so.2 on
i486 scratchbox2 targets doesn't really find correct libraries. This means
that only first level of dependencies were found but further levels were
missed in some cases.

To fix this, a recursive completely objdump based search for non-runnable
binaries was implemented.

This is slightly slower than ld-linux, but seems to work robustly
on i486 scratchbox2.

Signed-off-by: Kalle Jokiniemi <kalle.jokiniemi@jolla.com>

### CODE REVIEW

**Comment by kjokinie** [comment]
Tested this also (in addition to i486) on armv7hl target (Jolla Phone initrd creation) and it works with binaries used there as well.

**Comment by thp** [comment]
What about using `ldd` instead of `objdump`? Also, what about using `perl` or `python` instead of `sh` for doing the output parsing? E.g. [dpkg-shlibdeps](https://anonscm.debian.org/cgit/dpkg/dpkg.git/tree/scripts/dpkg-shlibdeps.pl) is written in Perl and presumably takes care of many corner cases. `dlopen()`-style dependencies aren't handled by it, though.

**Comment by kjokinie** [comment]
ldd "runs" the object in scratchbox2, which is the exact thing that is broken on i486 targets, as some things get executed from host side and others from the target and the paths point to weird places like /parentroot/srv/mer/sdk/targets/i486/usr/lib... I think this is because scratchbox just hasn't received that much testing for running i486 apps (as you can run them natively in most cases).

I guess "LD_TRACE_OBJECTS=1 <executable>" is pretty much == "ldd <executable>"..

I'm just not a perl or python person, sorry. But yes, I will happily use existing tool if there is a working one packaged in some mer/nemo repo. Is there some git repo for that dpkg tool?

**Comment by kjokinie** [comment]
Please disregard the stypid question about git repo :P

**Comment by thp** [comment]
If `dpkg-shlibdeps` is easy to use in the environment we have, I'd be in favor of using that. Otherwise, the tool we have here probably also does the job for the time being.

So if you have time to do some experimentation on `dpkg-shlibdeps` (basically seeing if it can be used easily without much effort), try that.

File: initfs/scripts/moslo-build.sh [hunk 1]
```diff
@@ -60,17 +60,43 @@ debug()
...
+                echo "$0: Error: Empty executable name!" 1>&2
+                exit 1
+        fi
+
+        if ! test -f $2; then
+                echo "$0: Error: No temporary list file provided!" 1>&2
                 exit 1
         fi
+
+        if cat $2 | grep -w $1; then
```
Comment (thp) [inline initfs/scripts/moslo-build.sh]: Can be rewritten as `if grep -w $1 $2; then`, saving one process?

File: initfs/scripts/moslo-build.sh [hunk 2]
```diff
@@ -60,17 +60,43 @@ debug()
...
+                # This library is already in the list, skipping.
+                return
+        else
+                # Clean up temp file from duplicates
+                SORTED_FILE=$(cat $2 | sort | uniq)
```
Comment (thp) [inline initfs/scripts/moslo-build.sh]: Can be rewritten as `SORTED_FILE=$(sort -u "$2")`

File: initfs/scripts/moslo-build.sh [hunk 3]
```diff
@@ -84,6 +110,20 @@ objdump_find_lib_path()
                         fi
                 done
         done
+
+        RECURSED_LIBS="$FOUND_LIBS"
+
+        for recursed_lib in $RECURSED_LIBS; do
+                if cat $2 | grep -w $recursed_lib; then
```
Comment (thp) [inline initfs/scripts/moslo-build.sh]: Can be rewritten as `if grep -q $recursed_lib $2; then`

File: initfs/scripts/moslo-build.sh [hunk 4]
```diff
@@ -84,6 +110,20 @@ objdump_find_lib_path()
...
+                        continue
+                else
+                        FOUND_LIBS="$FOUND_LIBS $(objdump_find_lib_paths $recursed_lib $2)"
+                fi
+        done
+
+        # Sort out duplicates to reduce amount of recursion.
+        FOUND_LIBS=$(echo $FOUND_LIBS | sed -e "s/ /\n/g" | sort | uniq)
```
Comment (thp) [inline initfs/scripts/moslo-build.sh]: `| sort | uniq` can be shortened to `| sort -u`

**Comment by thp** [comment]
Assuming the little `dpkg-shlibdeps` doesn't work out, I've reviewed this PR to take it in as-is. Added some comments. In general, for shell scripts, run [shellcheck](https://github.com/koalaman/shellcheck) and fix any quoting issues / ... it tells you about.

After the comments have been addressed, and shellcheck is happy with the script, LGTM.

**Comment by kjokinie** [comment]
It seems dpkg is not packaged for Mer. The standalone script needs some internal Dpkg.pm file (is that some perl module?) at least which would mean it needs to be packaged. I'm not too eager to start packaging dpkg for Mer.

File: initfs/scripts/moslo-build.sh (hunk 1 above)
Comment (kjokinie) [inline initfs/scripts/moslo-build.sh]: Hah :)

I've been complained about my cat|grepping before, but this is my personal fixation stemming from 1st university course about usage of Unix. "in unix we have this thing called pipes.. so this is how you search something from a file cat | grep...". I think Janne Lääkkö (the original author of mosle-build) is also from Oulu university (the cat | sort | uniq copy-paste) ;)

File: initfs/scripts/moslo-build.sh (hunk 2 above)
Comment (kjokinie) [inline initfs/scripts/moslo-build.sh]: Ok, the sort -u looks cleaner. I'll change those.

**Comment by thp** [comment]
Yep, so let's drop the dpkg idea and go forward with merging this after the comments and shellcheck have been dealt with :)

**Comment by kjokinie** [comment]
Would you like me to go through rest of shell scripts in Mer/Nemo/Sailfish as well with shellcheck? The script is not written with that coding style. I don't see point doing it for a single patch when rest of the code is not using it.

**Comment by kjokinie** [comment]
Updated version 2.

File: initfs/scripts/moslo-build.sh (hunk 1 above)
Comment (thp) [inline initfs/scripts/moslo-build.sh]: Do you plan to still change this or leave it like that?

File: initfs/scripts/moslo-build.sh (hunk 3 above)
Comment (thp) [inline initfs/scripts/moslo-build.sh]: Same here, will you fix this or leave it like that?

**Comment by thp** [comment]
Just make sure your additions and changes are `shellcheck`-clean. Somebody at some point should go and shellcheck the whole project and fix stuff. Obviously not as part of this PR ;)

**Comment by kjokinie** [comment]
Did you see my reply? There is no rule that every script has to be written with shellcheck. Is there a functional issue with the patch?

**Comment by thp** [comment]
It's not a rule, it's good practice, and avoids issues/bugs/security issues down the road. No functional issues with the patch, just some things that could be done cleaner / more efficient.

LGTM if you think it's good enough for your quality standards.

**Pull request merged**